### DISCUSSION TITLE
Error h2o wave container

### DISCUSSION THREAD

**Discussion by rmfjd78** [h2oai/wave#1310] [Q&A]
In a Container is Wave Server neded or i can just make pip install h2o_wave

I'm creating a container with a h2owave app and i'm getting the folowing error:
File "/opt/venv/lib/python3.9/site-packages/uvicorn/protocols/http/h11_impl.py", line 389, in run_asgi
result = await app(self.scope, self.receive, self.send)
File "/opt/venv/lib/python3.9/site-packages/uvicorn/middleware/proxy_headers.py", line 45, in __call__
return await self.app(scope, receive, send)
File "/opt/venv/lib/python3.9/site-packages/uvicorn/middleware/asgi2.py", line 7, in __call__
await instance(receive, send)

here are:
```
---------------------requirements.txt
certifi==2021.10.8
charset-normalizer==2.0.12
click==7.1.2
cycler==0.11.0
DateTime==4.4
fonttools==4.31.2
future==0.18.2
h11==0.13.0
h2o==3.36.0.3
h2o-wave==0.20.0
httpcore==0.12.3
httpx==0.16.1
idna==3.3
kiwisolver==1.4.1
matplotlib==3.5.1
missingno==0.5.1
numpy==1.22.3
packaging==21.3
pandas==1.4.1
Pillow==9.0.1
pydantic==1.9.0
pyparsing==3.0.7
python-dateutil==2.8.2
pytz==2022.1
requests==2.27.1
rfc3986==1.5.0
scipy==1.8.0
seaborn==0.11.2
six==1.16.0
sniffio==1.2.0
starlette==0.13.8
tabulate==0.8.9
typing==3.7.4.3
typing_extensions==4.1.1
urllib3==1.26.9
uuid==1.30
uvicorn==0.12.2
zope.interface==5.4.0
```
---------------------dockerfile:
```dockerfile
FROM python:3.9-slim AS builder

#ENV http_proxy="http://#{Proxy}#"
#ENV https_proxy="http://#{Proxy}#"


RUN apt-get update && \
   apt-get install -y gcc g++ ca-certificates gnupg gnupg1 gnupg2 && \
   apt-get install -y --no-install-recommends apt-utils \
   libpq-dev \
   uuid \
   poppler-utils && \
   rm -rf /var/lib/apt/lists/*

RUN python -m venv /opt/venv
ENV PATH="/opt/venv/bin:$PATH"

ADD ./requirements.txt .

RUN pip install --upgrade pip && \
    pip install --no-cache-dir -r requirements.txt && \
    find /opt/venv \
        \( -type d -a -name test -o -name tests -o -name '__pycache__' \) \
        -o \( -type f -a -name '*.pyc' -o -name '*.pyo' \) \
        -exec rm -rf '{}' \;


FROM python:3.9-slim

COPY --from=builder /opt/venv /opt/venv
ENV PATH="/opt/venv/bin:$PATH"

#ENV http_proxy="http://#{Proxy}#"
#ENV https_proxy="http://#{Proxy}#"

RUN apt-get update && \
   apt-get install -y --no-install-recommends \
   gunicorn3 && \
   rm -rf /var/lib/apt/lists/*

ADD ./App /usr/app
WORKDIR /usr/app

CMD python -m uvicorn main:app --host 0.0.0.0 --port 10101
```
```dockerfile
---------------------Docker Compse
version: '3.9'
services:
  utopia:
    image: "utopia"
    ports:
      - "10101:10101"
    deploy:
      resources:
        limits:
          cpus: "2"
          memory: 600M
      restart_policy:
        condition: on-failure
      mode: replicated
      replicas: 1
```

**Comment by mturoci** [comment]
Hi @rmfjd78, although you did not post the whole error stack trace, by looking at the dockerfile, I assume the problem is your app cannot connect to the Wave server (waved) since I do not see starting it anywhere. You have 2 options how to go about this:

* for Wave > 0.20, waved is packaged together with py distribution so you should be able to simple use `wave run your-app` within `CMD`. This is the simplest, least flexible solution.
* Another solution is to run waved and wave app separately (the only option for Wave < 0.19). This approach is a bit more complex, but allows for more config flexibility. From what you have posted, I do not see waved startup. See https://github.com/h2oai/wave/pull/1234 for guide on how to dockerize Wave.

**Reply by rmfjd78** [reply]
Since i’m using the lates version of the h2o_wave in my container i have the following problem when i start the container and try to access my App in the browser the page doesn’t load. Do i need to configure anything in the container (proxy, ….) or the wave command needs extra parameters?

**Reply by mturoci** [reply]
Try this:

```dockerfile
# Get base image with python installed.
FROM python:3.8
# Copy over everything.
COPY . .
# Install needed dependencies.
RUN pip install -r requirements.txt
# Run the app.
CMD ["wave", "run", "--no-reload", "app.py"]
```

**Comment by rmfjd78** [comment]
Since i’m using the lates version of the h2o_wave in my container i have
the following problem when i start the container and try to access my App
in the browser the page doesn’t load. Do i need to configure anything in
the container (proxy, ….) or the wave command needs extra parameters?

mturoci ***@***.***> escreveu em ter., 29/03/2022 às 08:07 :
